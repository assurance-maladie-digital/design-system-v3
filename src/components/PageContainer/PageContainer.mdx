import {Controls, Canvas, Meta, Source} from '@storybook/blocks';

import * as PageContainerStories from './PageContainer.stories';

<Meta of={PageContainerStories} />

# PageContainer

Le composant `PageContainer` est un conteneur transparent utilis√© pour afficher une page.

<Canvas of={PageContainerStories.Default}  />

# API

<Controls of={PageContainerStories.Default} />

# Exemple d'utilisation

<Source dark code={`
<script setup lang="ts">
import PageContainer from '@cnamts/synapse'
</script>

<template>
    <PageContainer color="primary">
      Contenu de la page
    </PageContainer>
</template>
`} />