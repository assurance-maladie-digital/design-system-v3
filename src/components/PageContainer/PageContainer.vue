<script setup lang="ts">
	import { computed } from 'vue'
	import { useDisplay } from 'vuetify'

	const props = withDefaults(defineProps<{
		size?: 'xl' | 'l' | 'm' | 's'
		spacing?: 'xs' | 'sm' | 'md' | 'lg' | 'xl'
		color?: string
	}>(), {
		size: 'xl',
		spacing: undefined,
		color: 'transparent',
	})

	const display = useDisplay()

	const spacingMapping: Record<string, string> = {
		xs: 'px-0',
		sm: 'px-4',
		md: 'px-8',
		lg: 'px-14',
		xl: 'px-14',
	}

	const spacingX = spacingMapping[display.name.value]

	const spacingClass = computed(() => {
		if (props.spacing) {
			return `py-10 ${spacingMapping[props.spacing]}`
		}

		return `py-10 ${spacingX}`
	})

	const sizeMapping: Record<string, number> = {
		xl: 1440,
		l: 960,
		m: 800,
		s: 600,
	}

	const containerSize = computed(() => {
		return sizeMapping[props.size]
	})

	defineExpose({
		spacingClass,
		containerSize,
	})
</script>

<template>
	<div :class="[spacingClass, 'vd-page-container d-flex justify-center']">
		<VSheet
			:width="containerSize"
			:color="color"
		>
			<slot />
		</VSheet>
	</div>
</template>

<style lang="scss" scoped>
.vd-page-container {
  flex: 1;
  max-width: 1712px;
}
</style>
