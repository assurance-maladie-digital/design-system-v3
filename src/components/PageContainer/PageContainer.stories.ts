import type { Meta, StoryObj } from '@storybook/vue3'
import PageContainer from './PageContainer.vue'
import { VCard } from 'vuetify/components'

const meta = {
	title: 'Components/PageContainer',
	component: PageContainer,
	parameters: {
		layout: 'fullscreen',
		controls: { exclude: ['spacingClass', 'containerSize'] },
	},
	argTypes: {
		size: {
			options: ['xl', 'l', 'm', 's'],
			control: { type: 'select' },
			default: 'xl',
		},
		spacing: {
			options: ['xs', 'sm', 'md', 'lg', 'xl'],
			control: { type: 'select' },
			default: undefined,
		},
		color: {
			options: ['transparent', 'primary', 'secondary', 'accent', 'error', 'info', 'success', 'warning'],
			control: { type: 'select' },
			default: 'transparent',
		},
	},
} as Meta<typeof PageContainer>

export default meta

type Story = StoryObj<typeof meta>

export const Default: Story = {
	parameters: {
		sourceCode: [
			{
				name: 'Template',
				code: `<template>
	<PageContainer>
		Contenu de la page
	</PageContainer>
</template>
				`,
			},
			{
				name: 'Script',
				code: `<script setup lang="ts">
	import PageContainer from '@cnamts/synapse'
</script>
				`,
			},
		],
	},
	args: {
		default: 'Contenu de la page',
	},
	render: (args) => {
		return {
			components: { PageContainer },
			setup() {
				return { args }
			},
			template: `
                <PageContainer :size="args.size" :spacing="args.spacing" :color="args.color">
					{{ args.default }}
                </PageContainer>
            `,
		}
	},
}

export const Size: Story = {
	parameters: {
		sourceCode: [
			{
				name: 'Template',
				code: `<template>
	<PageContainer size="s">
		Contenu de la page
	</PageContainer>
</template>
				`,
			},
			{
				name: 'Script',
				code: `<script setup lang="ts">
	import PageContainer from '@cnamts/synapse'
</script>
				`,
			},
		],
	},
	args: {
		default: 'Contenu de la page',
		size: 's',
	},
	render: (args) => {
		return {
			components: { PageContainer },
			setup() {
				return { args }
			},
			template: `
				<PageContainer :size="args.size" :spacing="args.spacing" :color="args.color">
					{{ args.default }}
				</PageContainer>
			`,
		}
	},
}

export const Color: Story = {
	parameters: {
		sourceCode: [
			{
				name: 'Template',
				code: `<template>
	<PageContainer color="primary">
		<div class="pa-8">
			Contenu de la page
		</div>
	</PageContainer>
</template>
				`,
			},
			{
				name: 'Script',
				code: `<script setup lang="ts">
	import PageContainer from '@cnamts/synapse'
</script>
				`,
			},
		],
	},
	args: {
		default: 'Contenu de la page',
		color: 'primary',
	},
	render: (args) => {
		return {
			components: { PageContainer },
			setup() {
				return { args }
			},
			template: `
				<PageContainer :size="args.size" :spacing="args.spacing" :color="args.color">
					<div class="pa-8">
						{{ args.default }}
					</div>
				</PageContainer>
			`,
		}
	},
}

export const Card: Story = {
	parameters: {
		sourceCode: [
			{
				name: 'Template',
				code: `<template>
	<PageContainer>
		<VCard class="pa-8">
			Contenu de la page
		</VCard>
	</PageContainer>
</template>
				`,
			},
			{
				name: 'Script',
				code: `<script setup lang="ts">
	import PageContainer from '@cnamts/synapse'
	import { VCard } from 'vuetify/components'
</script>
				`,
			},
		],
	},
	args: {
		default: 'Contenu de la page',
	},
	render: (args) => {
		return {
			components: { PageContainer, VCard },
			setup() {
				return { args }
			},
			template: `
				<PageContainer :size="args.size" :spacing="args.spacing" :color="args.color">
					<VCard class="pa-8">
						{{ args.default }}
					</VCard>
				</PageContainer>
			`,
		}
	},
}
