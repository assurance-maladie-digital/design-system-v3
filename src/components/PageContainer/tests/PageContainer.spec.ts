import { describe, it, expect } from 'vitest'
import { mount } from '@vue/test-utils'
import { vuetify } from '@tests/unit/setup'

import PageContainer from '../PageContainer.vue'

describe('PageContainer', () => {
    it('render correctly', async () => {
        const wrapper = mount(PageContainer, {
            global: {
                plugins: [vuetify],
            },
        })

        expect(wrapper.html()).toMatchSnapshot()
    })

    it('render correctly with slot', async () => {
        const wrapper = mount(PageContainer, {
            slots: {
                default: 'slot content',
            },
            global: {
                plugins: [vuetify],
            },
        })

        expect(wrapper.html()).toContain('slot content')
    })

    it('render correctly with spacing class', async () => {
        const wrapper = mount(PageContainer, {
            props: {
                spacing: 'ma-4',
            },
            global: {
                plugins: [vuetify],
            },
        })

        expect(wrapper.vm.spacingClass).toBe('ma-4')
    })

    it('containerSize computed property', async () => {
        const wrapper = mount(PageContainer, {
            props: {
               size: 'l',
            },
            global: {
                plugins: [vuetify],
            },
        })

        expect(wrapper.vm.containerSize).toBe(960)
    })
})

