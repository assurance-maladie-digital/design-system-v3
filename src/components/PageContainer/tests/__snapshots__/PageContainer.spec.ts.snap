// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`PageContainer > render correctly 1`] = `
"<div data-v-fb0b02fd="" class="py-10 px-8 vd-page-container d-flex justify-center">
  <p data-v-fb0b02fd="">Container size: 1712</p>
  <div data-v-fb0b02fd="" class="v-sheet v-theme--light bg-transparent" style="width: 1712px;"></div>
</div>"
`;
